Various updates to subsys/mgmt/osdp

sidcha opened the pull request:
This PR brings in various updates such as bug fixes, cleanups and new features to subsys/mgmt/osdp in both control panel and peripheral device modes. These changes have been tested with a st-nucleo-g474re board with secure channel enabled.

Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.

sidcha committed:
boards: arm: stm32: Add dt-chosen uart for sybsys/mgmt/osdp

Add dt-chosen entry osdp-uart for nucleo_g474re board as it is an ideal
candidate for trying out all features of sybsys/mgmt/osdp.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
samples: mgmt/osdp: Select ENTROPY_GENERATOR from prj.conf

ENTROPY_GENERATOR is needed for OSDP Secure Channel capability. Now
that Secure Channel patches are merged, enabling it by default.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Cleanup log messages and return codes

The log lines in CP and PD had a prefix such as "CP: " and "PD: " that
does not add too much value as a given device an either be CP or PD
only. This patch removes those and enhances some other log lines while
at it.

It also adds a enum for return values throughout the module to improve
code quality.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Flush RX buffer before sending data

Partial packets in the RX buffers cause the subsequent packet to be
treated as malformed. The RX buffer can have partial data if the sender
is too slow in sending the packet of if there is an interruption in
transmission mid-way.

To avoid any issues due to such partials, flush the uart channel before
sending the command/response.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Rename cmd_data as ephemeral_data

Since cmd_data member is used by both commands and events to store the
contents of current transaction, rename it to ephemeral_data which
better reflects the purpose of the variable.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Rename pd offset as index

CP has an array of PDs and pd->offset was the position of the PD in CP's
list. Since offset has many meanings, rename it to pd->idx.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Add capability checks on incoming commands

OSDP compliant devices communicate their capabilities and discover what
their peer can and cannot do. Right now, PD advertises these capabilities
and expects CP to honor them. Although this is not known to cause any
issues, it is not desirable to allow such accesses.

Add a check of incoming commands to to validate that the corresponding
capability was enabled and advertised.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Add support for key press and card read events

Now that we have the necessary infrastructure to collect events from PD
apps, we can use them to translate it to OSDP packet sequence for card
reads and key press events.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Add inline methods for flag checks

The raw, flags check has become a bit excessive and has begun to affect
code readability. Provide inline functions for those accesses that are
frequent. Also, get rid of `struct osdp_cp` as it can be fully represented
by `struct osdp` itself.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Rework MARK byte handling

OSDP specification section 5.7 states that a transmitting device has to
drive the transmission line to a marking state for a period of one char
in the current baud rate. This can be achieved by sending 0xFF. Since
this is not mentioned in the packet structure definition, many commercial
implementations of OSDP out in the wild do not send/expect this byte.

To work with such non-conforming devices, we will try to be as flexible
as possible in the PD: send mark byte only if the other side sent one. In
case of CP, we have no option but to send the mark byte to be as close
to the specification as possible. If a particular use case needs the CP
to not send it, we will provide a Kconfig option to disable it.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Catch out-of-order SC packets and fail

During handshake, only certain types of secure block types (<= SCS_14)
are allowed. A rouge CP/PD can try to bypass the handshake by directly
sending a secure block type ahead of the sequence and gain a a secure
channel. Fix this by adding a check in packet decode time.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: cp: Move set PD online status to a method

When a PD is set online, we need to do other actions along with it. To
make this easier in future, move cp_set_state(pd, OSDP_CP_STATE_ONLINE)
to it's own method cp_set_online(pd).

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: cp: Split build_packet and send_command

The method send_command was building the packet and also sending it.
This is a problem because even command build failures are reported as
command send failures; although the former can be mitigated by changing
kconfig parameters depending on use cases.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Move packet checks out of decode_packet

To handle multiple packets in RX buffer, we need to perform checks on
the buffer first and determine the length of one packet and extact only
that many bytes from the RX buffer instead of pulling everything at
once and failing if there are bytes from the next packet in the buffer.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Rework state machine to handle multiple packets

After the check_packet() split from decode_packet() for handling multiple
packets in the RX buffer, PD state machine needed to me modified to
start handling it correctly. While at it, simplify the states as PD does
not have as much state anyways.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: api: Add support for command complete callback

Add command complete callback so app can subscribe to osdp command or
event completion events. This can be used to perform post command
actions such as changing the baud rate of the underlying channel after a
COMSET command was acknowledged/issued by a peer.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: cp: Rework CP state management

With changes to the phy layer and the way we build and dispatch
commands, we can now simplify some parts of CP state machine. This patch
builds on previous patches to reduce some state transitions and fixes
some bugs in those paths.

While at it, also implement an exponential back-off for PD offline after
a timeout. This is useful when there is a transient PD timeout but the
PD itself is not totally dead.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: More error handling and other cleanups

cp_build_command returns the length of the built packet and early
returns for errors. This means there is no need to keep track of return
code in a separate variable -- if we reached the end, we succeeded.

Along with this change, some more changes to enhance maintainability and
readability are also done.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Fix device capabilities report

Do not check or send the first entry in the pd->cp[] device capability
table which is for function code 0 which is not a defined function code.

Signed-off-by: David Vucich <dave@alcatraz.ai>
Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: sc: Restructure loop to avoid a bunch of -1s

Hoisting a `length--` helps simplify a lot of code which has ad-hoc -1s
in a  bunch of places. Also, fix some formatting issues and remove
unnecessary log lines.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Rework secure channel key management

OSDP command KEYSET is used to set the secure channel base key for all
connected PDs. This key is then used to derive the session keys for each
secure channel session. When the app wants to set the this key, it has
to issue a command and then both the CP/PD has to be notified of this
change so they can store this key in a non-volatile medium for future
operations across power cycles.

The current implementation of OSDP had limited support for key
management. This patch adds all the bells and whistles needed to handle
keyset command/event in the CP/PD application.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Extend get_sc_status_mask to handle more than 32 PDs

Today, the API to get secure channel status returns a uint32_t which
means it can handle only upto 32 PDs at most. Extent this API to take a
uint8_t pointer allocated by caller so we can handle all PDs that a CP
can support (which is 126 today).

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Add API to get status of all PDs

Add an API to get the status of all configured PDs as a bitmask. The
bit position indicated the PD number and the bit value indicates if the
PD is online.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Fix doxygen comment formatting in osdp.h

Fix some alignment issues in doxygen comments in osdp.h and add section
markers to make it easy to jump to different sections of the file.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Allow non-conformant, 0 length, encrypted data blocks

If command or reply has no data, PD "should" use secure message types
SCS_15 or SCS_16. But some PD seem to not implement this correctly. We
will be tolerant towards those faulty implementations.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Allow 0 length, encrypted data blocks

Some PDs sends an ACK for a KEYSET, this because it has discarded the
secure channel session keys in favour of the new key we sent and hence
this packet may reach us in plain text. This is not as stated in the
specification, if command or reply has no data, PD should use secure
message types SCS_15 or SCS_16.

Since a lot of commercial PDs get it wrong, we must also discard our
secure session to work with them.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Add support for manufacturer command/reply

OSDP specifies manufacturer specific commands and replies that can be
used my the PD app to defined custom data transport over OSDP leveraging
the secure channel features.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
MAINTAINERS: Mark subsys/mgmt/osdp as maintained

subsys/mgmt/osdp _has_ a dedicated maintainer who will rush to address
any issues identified :)

Also, add @adakus and @r2r0 as collaborators since they are using this
subsystem in production and have agreed to review, test, and report issues
for this subsystem in future.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -1220,7 +1220,7 @@ MCU Manager:
         - "area: mcumgr"
 
 OSDP:
-    status: odd fixes
+    status: maintained
     collaborators:
         - sidcha
```
> subsys/mgmt/osdp _has_ a dedicated maintainer who will rush to address
any issues identified :).

In that case, please add yourself as a maintainer.

sidcha replied:
Done.

adakus commented on include/zephyr/mgmt/osdp.h:
```diff
@@ -202,38 +215,251 @@ struct osdp_cmd {
+ * @param ctx OSDP context
+ * @param cb Callback to be invoked when a command is completed.
+ */
+void osdp_set_command_complete_callback(osdp_command_complete_callback_t cb);
```
I guess this one should be outside of #ifdef CONFIG_OSDP_SC_ENABLED ... #endIf

sidcha replied:
Thanks for the review, I disabled CONFIG_OSDP_SC_ENABLED and fixed this and other issues that were caused due to code leaks our side this switch.

sidcha replied:
Also, CONFIG_OSDP_SC_ENABLED situation has worsened now more than ever. So I'll add another patch to reduce the number of  ifdefs we use to disable secure channel.

adakus commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -678,77 +939,90 @@ static int pd_receve_packet(struct osdp_pd *pd)
+	 * If secure channel is established, we need to make sure that
+	 * the session is valid before accepting a command.
+	 */
+	if (sc_is_active(pd) &&
```
For CONFIG_OSDP_SC_ENABLED == 0 struct member  'sc_tstamp' is not visible

sidcha replied:
Fixed (see above).

adakus commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -432,53 +639,78 @@ static int pd_build_reply(struct osdp_pd *pd, uint8_t *buf, int max_len)
-		}
+		ASSERT_BUF_LEN(REPLY_RSTATR_LEN);
 		buf[len++] = pd->reply_id;
 		buf[len++] = ISSET_FLAG(pd, PD_FLAG_R_TAMPER);
```
According to spec this status can take values in range: 0-2

sidcha replied:
Right. But it is the current limitation of the implementation and this patch-set is not doing anything about it. We can take this up as a separate task once this PR is merged.

sidcha committed:
mgmt/osdp: Reduce ifdefs needed to disable secure channel

The secure channel feature of OSDP is guarded by kconfig switch
CONFIG_OSDP_SC_ENABLED. Since SC was added later in the development
cycle, it made sense to keep those changes inside a macro switch to make
sure we weren't regressing. But now after all the cleanups, the number
of places we need to add ifdefs to disable secure channel has increased
significantly; to a point where it is hindering readability.

This patch minimizes the number of places we need to have such ifdefs by
doing two things:
  * having all SC data members even when SC is disabled. So ~200 bytes per
    PD of RAM is wasted.
  * moving check to a static inline method sc_is_enabled() which most
    compilers should be able to optimize for us.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Add length checks for commands and replies

For all commands and replies, the buffer length needed to build or the
length of data needed to decode needs to be checked and asserted. Right now
we do this by ad-hoc if-s. Add macros that do this at a common location.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Add support for event delivery and notifications

The CP app sends PD a "command" and the PD responds to it. Some times,
the PD has something that it wants to tell the PD which it does so in
response to POLL command. Both CP and PD apps need a way to exchange
these info over the OSDP bus. To archive this we will introduce what are
called "events" that allow the PD app to enqueue and CP app to get
notified.

This is analogous to the incumbent "commands" abstraction where, the CP
app enqueues a command and the PD app gets notified of it.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha commented:
I think the 4 checkpatch issues due to the following are valid exceptions to that rule.

```
-:1339: WARNING:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
#1339: FILE: subsys/mgmt/osdp/src/osdp_cp.c:133:
+	#define ASSERT_BUF_LEN(need)                                         \
+		if (max_len < need) {                                        \
+			LOG_ERR("OOM at build CMD(%02x) - have:%d, need:%d", \
+				pd->cmd_id, max_len, need);                  \
+			return OSDP_CP_ERR_GENERIC;                          \
+		}
```

gmarull commented:
> I think the 4 checkpatch issues due to the following are valid exceptions to that rule.
> 
> ```
> -:1339: WARNING:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
> #1339: FILE: subsys/mgmt/osdp/src/osdp_cp.c:133:
> +	#define ASSERT_BUF_LEN(need)                                         \
> +		if (max_len < need) {                                        \
> +			LOG_ERR("OOM at build CMD(%02x) - have:%d, need:%d", \
> +				pd->cmd_id, max_len, need);                  \
> +			return OSDP_CP_ERR_GENERIC;                          \
> +		}
> ```

I agree with checkpatch here, a macro named "ASSERT" should just assert, not obfuscate logic.

sidcha commented:
> I agree with checkpatch here, a macro named "ASSERT" should just assert, not obfuscate logic

The name is misleading then, I can change it to something more appropriate. These conditions can happen at run time due to elevated activity or due to a malformed command/response; if so we send out an error response.

sidcha commented:
@gmarull, I've removed the use of the macro by converting half of them into plan asserts (where we control things) and the other half to literal code substitutions due to lack of other options.

Now all tests are passing!

sidcha committed:
mgmt/osdp: Fix packet trace logging

OSDP raw packet trace logging was broken on the PD side because it was not
accounting for secure message block when secure channel was enabled. On top
of that, after the introduction of CONFIG_OSDP_SKIP_MARK_BYTE, it was
broken in CP too. Fix both these issues and also enhance the osdp_dump
method to accept variadic args to avoid a LOG_DBG prefix.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

carlescufi commented:
@adakus please take another look

sidcha commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -105,279 +116,523 @@ static struct osdp_pd_cap osdp_pd_cap[] = {
 		if (len != CMD_OSTAT_DATA_LEN) {
 			break;
 		}
+		if (!pd_cmd_cap_ok(pd, NULL)) {
```
Oops! Some rebase artefact. Fixed, will be fixed in next version.

sidcha commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -105,279 +116,523 @@ static struct osdp_pd_cap osdp_pd_cap[] = {
+	return reply_code;
+}
+
+static int pd_cmd_cap_ok(struct osdp_pd *pd, struct osdp_cmd *cmd)
```
Added checks on cmd where it is used. Thanks.

adakus commented:
Hi,
I have bug in your implementation

***@***.***

When SC is not enabled then:

***@***.***

Only one item is copied – for second item function code is equal to 0.
So must be fixed – otherwise other capabilities are not set using config defines
Best regards,
Adam

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: Siddharth ***@***.***>
Sent: Thursday, July 7, 2022 10:40 PM
To: ***@***.***>
Cc: Kusiek, ***@***.***>; ***@***.***>
Subject: [EXT] Re: [zephyrproject-rtos/zephyr] Various updates to subsys/mgmt/osdp (PR #46449)

*** Please use caution this is an externally originating email. ***

@sidcha commented on this pull request.


In subsys/mgmt/osdp/src/osdp_pd.c<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fpull%2F46449%23discussion_r916263444&data=05%7C01%7Cadam.kusiek1%40hidglobal.com%7C634b266e000746cb79b908da6058dc1c%7Cf0bdc1c951484f86ac40edd976e1814c%7C0%7C0%7C637928232101044637%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4ZNAF1Ea8Xp4cdYZ0fsUIO6e4ksuKZqaw0cri%2BrTNCI%3D&reserved=0>:

> +     case OSDP_EVENT_KEYPRESS:

+              reply_code = REPLY_KEYPPAD;

+              break;

+       default:

+              LOG_ERR("Unknown event type %d", event->type);

+              break;

+       }

+       if (reply_code == 0) {

+              /* POLL command cannot fail even when there are errors here */

+              return REPLY_ACK;

+       }

+       memcpy(pd->ephemeral_data, event, sizeof(struct osdp_event));

+       return reply_code;

+}

+

+static int pd_cmd_cap_ok(struct osdp_pd *pd, struct osdp_cmd *cmd)

Added checks on cmd where it is used. Thanks.

—
Reply to this email directly, view it on GitHub<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzephyrproject-rtos%2Fzephyr%2Fpull%2F46449%23discussion_r916263444&data=05%7C01%7Cadam.kusiek1%40hidglobal.com%7C634b266e000746cb79b908da6058dc1c%7Cf0bdc1c951484f86ac40edd976e1814c%7C0%7C0%7C637928232101200900%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Vy5kVjkogzWr8n0MYEmp6BCDiv7MoA8H2MdTROv2R2w%3D&reserved=0>, or unsubscribe<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAZKOCFIPT7V5HWKJKWKD7B3VS46B5ANCNFSM5YO4VQ6Q&data=05%7C01%7Cadam.kusiek1%40hidglobal.com%7C634b266e000746cb79b908da6058dc1c%7Cf0bdc1c951484f86ac40edd976e1814c%7C0%7C0%7C637928232101200900%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Xh37HW%2BD0cU6mWJqKzyeByaHUdh35AKvCWrVuvh22BQ%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

sidcha commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -57,8 +68,8 @@ static struct osdp_pd_cap osdp_pd_cap[] = {
 		0, /* N/A */
 	},
 	{
-		OSDP_PD_CAP_COMMUNICATION_SECURITY,
 #ifdef CONFIG_OSDP_SC_ENABLED
+		OSDP_PD_CAP_COMMUNICATION_SECURITY,
```
This hunk should not be here, will be removed in next revision. Thanks @adakus.

sidcha commented:
Can we have some eyes on this one please :)

cc: @gmarull, @carlescufi, @adakus, @stephanosio.

sidcha commented:
@galak, can you look at this PR as well :)

r2r0 commented:
@MaureenHelm I ask you very kindly as a code owner - could you please consider reviewing this PR or discuss topic who is in charge to do that?
It looks that it is somewhat catch 22 - no owner then no review then no owner assignement then no review ...
We (as HID company) use this code in production environment and it will help us a lot to have it in main branch.

stephanosio reviewed (dismissed):
I haven't done an in-depth review of all the changes, but they look reasonable at first glance.

adakus commented on include/zephyr/mgmt/osdp.h:
```diff
@@ -202,40 +215,248 @@ struct osdp_cmd {
+ * @param length Length of manufacturer data in bytes (optional)
+ * @param data manufacturer data of `length` bytes (optional)
+ */
+struct osdp_event_mfgrep {
```
Relevant events/cmds should have the same size of members. 
This comes out from the fact that event can be used as response on POLL cmd. 
In this situation event (provided in ephemeral data) is directly casted to relevant command struct so memory layout of cmd/event structs should be the same.

sidcha replied:
Thanks for the review. In PD, `CMD_MFG` copies the response from the app as `struct osdp_cmd` into ephemeral_data ([here](https://github.com/goToMain/zephyr/blob/osdp/subsys/mgmt/osdp/src/osdp_pd.c#L499)) and in `REPLY_MFGREP`, we cast it back to `struct osdp_cmd`.

So there is no problem with that code as it is now but still, I agree it can be a problem in future if we want to allow PDs to report unsolicited MFGREPs. I'll make `osdp_event_mfgrep` look like `osdp_cmd_mfg`

MaureenHelm commented:
> Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.

It's a lot of changes to review in one PR. You'll have better success getting reviews if you submit smaller batches of commits in separate PRs.

I've also added the `dev-review` label to put this PR on the meeting agenda for this week and find other stakeholders that can help review your changes. You're invited to join the meeting:
https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

sidcha commented:
> > Although this PR does way more than any single PR should be doing, I've taken care to split these changes across 30 commits to make them more digestible.
> 
> It's a lot of changes to review in one PR. You'll have better success getting reviews if you submit smaller batches of commits in separate PRs.

I know, I'm sorry. But I can explain :) 

> I've also added the `dev-review` label to put this PR on the meeting agenda for this week and find other stakeholders that can help review your changes. You're invited to join the meeting: https://github.com/zephyrproject-rtos/zephyr/wiki/Zephyr-Committee-and-Working-Group-Meetings#zephyr-dev-meeting

Thank you, I'll join the call.

sidcha commented:
@adakus, change you suggested is incorporated. Can you take a look again and approve this PR please?

@stephanosio, after your last review only a small fixup (above) was applied. Please take another look.

sidcha commented:
@r2r0, please take another look.

sidcha commented:
Thank you @adakus  and @stephanosio

sidcha commented:
@MaureenHelm we have 2 reviews now :)

MaureenHelm reviewed (changes requested):
Thank you for organizing and documenting your changes in a nice series of small atomic commits. However, public API and incorrect copyright changes are buried under a lot of implementation cleanup changes. I don't see anything controversial except for the copyright, but I'd like the public API changes submitted in a separate PR to give them a dedicated review.

sidcha commented on include/zephyr/mgmt/osdp.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020 Siddharth Chandrasekaran <siddharth@embedjournal.com>
+ * Copyright (c) 2020 Siddharth Chandrasekaran <sidcha.dev@gmail.com>
```
Having them both would defat the point (adding more confusion). I'll drop this change in the next revision.

stephanosio replied:
IMHO, this is not really a problem. This is not a "copyright" change; it is just updating an outdated email address.

sidcha replied:
@MaureenHelm can should I drop or keep this change?

sidcha commented on include/zephyr/mgmt/osdp.h:
```diff
@@ -211,24 +201,163 @@ struct osdp_cmd {
+ * @retval 0 on success
+ * @retval -1 on failure
+ */
+int osdp_pd_notify_event(struct osdp_event *event);
 
 #else /* CONFIG_OSDP_MODE_PD */
 
-int osdp_cp_set_callback_key_press(
-	int (*cb)(int address, uint8_t key));
-int osdp_cp_set_callback_card_read(
-	int (*cb)(int address, int format, uint8_t *data, int len));
```
There are currently 8 commits that can touch osdp.h:
```
23e778e38a mgmt/osdp: Reduce ifdefs needed to disable secure channel
c0177e7a58 mgmt/osdp: Change email address to gmail
2b38d1f370 mgmt/osdp: Add support for manufacturer command/reply
0d8ef2e529 mgmt/osdp: Fix doxygen comment formatting in osdp.h
03f63a30c8 mgmt/osdp: Add API to get status of all PDs
79ff7fd9f7 mgmt/osdp: Extend get_sc_status_mask to handle more than 32 PDs
17fa794d13 mgmt/osdp: api: Add support for command complete callback
1fe9943081 mgmt/osdp: Add support for event delivery and notifications
```
I don't know if they can be cleanly separated (some patches lead into the header medication patch). If this cannot be done, I'll just split the entire series into smaller PRs as suggested in dev-meeting and close this one.

sidcha commented:
@MaureenHelm I don't think splitting this PR is of much help; much let me explain. I just created #49302 with one logical block "introducing events: (it has an API change).  The rest of the 4 blocks ([2/5](https://github.com/goToMain/zephyr/tree/osdp_pd_cleanups), [3/5](https://github.com/goToMain/zephyr/tree/osdp_cp_cleanups), [4/5](https://github.com/goToMain/zephyr/tree/osdp_api_cleanups), [5/5](https://github.com/goToMain/zephyr/tree/osdp_misc_fixes)) depend on their predecessors' changes so they cannot be based on the main branch. If I create all 5 PRs at once, 5/5 would be identical to this one because of the way GitHub's UI presents the changes.

To review this PR is small chunks, I have to wait for 1/5 to be merged and then create 2/5 and so on. I would do that but is is just a lot of work for everyone :) So I suggest closing 1/5 and merging this one instead. What are your thoughts on this?

sidcha commented:
I've added @adakus and @r2r0 as collaborators for this subsystem after getting their consent over email.

str4t0m commented:
> @MaureenHelm I don't think splitting this PR is of much help;

I too prefer smaller PRs. If everything is crammed into a single PR errors are overlooked much more likely.
Also, sometimes collaborators/maintainers might have a few minutes available to squeeze in a PR review, but not enough to review a huge PR, leading to much quicker response times for smaller PRs.

In your current split of 5 PRs the fixes, API changes/additions, maintainer changes, and cleanups are still mixed together.
Why not create a PR with all/most of the cleanups and fixes first in a separate PR and API changes separately?
Also, I'd suggest to split the maintainer changes into a separate PR. There aren't any dependencies to other commits at all.
And in case the copyright changes block a merge, I'd suggest to leave them for now. Because they might even be merged after the merge window has closed. Although I am not completely sure about that.

sidcha commented:
@str4t0m,

> I too prefer smaller PRs. If everything is crammed into a single PR errors are overlooked much more likely. Also, sometimes collaborators/maintainers might have a few minutes available to squeeze in a PR review, but not enough to review a huge PR, leading to much quicker response times for smaller PRs.

I agree with this sentiment overall, truly. As I explained in the dev-review meeting, the reasons for such a large PRs were:
- this is a series of patches developed over a year in another port of this module. I wasn't made aware of any active users so cleanups and enhancements did not reach here (what is already merged is standalone and complete).
- this subsystem did not have dedicated reviewers/collaborators so getting changes merged was cumbersome. Now we have fixed that. Moving forward I'll make sure to send out smaller PRs

Overall, the request was to treat this as a special case and bear with me :) 

> [..] Why not create a PR with all/most of the cleanups and fixes first in a separate PR and API changes separately?

I tried that, It increases the rebase effort significantly. I might as well create-wait-merge-repeat 5 separate PRs.

> [..] Also, I'd suggest to split the maintainer changes into a separate PR.

This was easy. Created #49303 with maintainers and other ad-hoc changes.

> [..] And in case the copyright changes block a merge, I'd suggest to leave them for now.

Figured so, the change is question is already dropped in current version.

I'll close this PR in favour of the smaller ones since it isn't going anywhere.
